### PR TITLE
Switch PyTorch nightly to commit-based pin

### PR DESCRIPTION
**Pull request by huydhn**
This PR replaces the existing PyTorch nightly pin with a commit-based one where the PyTorch commit could be:

* A commit from main or viable/strict
* A commit from a WIP pull request
* or even a branch name

The main benefit of this change is to allow folks who are working on both PyTorch and ExecuTorch the flexibility of using the commit they want from PyTorch without the need to wait for that to become available on nightly build while leaving ExecuTorch CI in a potential broken state.

The changes here include:
* Remove `nightly.txt`
* PyTorch pinned commit is checkout and is built from source.
  * For Linux, the build happens during Docker image build, so there is no impact on CI duration.  From dev point of view, a PyTorch snapshot has always been installed at the specified commit
  * For MacOS, the build happens during the CI job.  This incurs additional build time, but I have added sccache to mitigate it.  The increase here is less than 5 minutes from my spot check, for example [before](https://github.com/pytorch/executorch/actions/runs/6948971118/job/18906105216) v.s. [after](https://github.com/pytorch/executorch/actions/runs/6951428944/job/18913383896)
* Audio, and vision will also need to be built from source (for compatibility with PyTorch C++ layer).  For simplicity, audio and vision commit pins will come from core 
* Install sccache and its openssl dependency
* Tweak the codebase here and there to make it work with building PyTorch from source

After this is ready, I will announce it to the team before landing the change.

### CODE REVIEW

**Comment by huydhn** [comment]
Caching is in used:

```
+ sccache --show-stats
Compile requests                   5628
Compile requests executed          4899
Cache hits                         4874
Cache hits (C/C++)                 4874
Cache misses                          1
Cache misses (C/C++)                  1
Cache timeouts                        0
Cache read errors                     0
Forced recaches                       0
Cache write errors                    0
Compilation failures                  2
Cache errors                         22
Cache errors (C/C++)                 22
Non-cacheable compilations            0
Non-cacheable calls                 509
Non-compilation calls               220
Unsupported compiler calls            0
Average cache write               0.091 s
Average compiler                  0.577 s
Average cache read hit            0.000 s
Failed distributed compilations       0

Non-cacheable reasons:
unknown source language             492
@                                     9
multiple input files                  6
Can't handle UnknownFlag arguments with -Xclang       2

Cache location                  s3, name: ossci-compiler-cache-circleci-v2, prefix: /executorch/
Version (client)                0.4.1
```

**Review by guangy10 — COMMENTED** [review]
> Audio, and vision will also need to be built from source (for compatibility with PyTorch C++ layer). For simplicity, audio and vision commit pins will come from core

I think this change makes more sense as these domain libs are required to be compatible with torch. The compatibility enforcement should hide from users, especailly after switching to commit-based pin, users will have harder time to figure the hashes of those domain libs that are compatibile with the pinned torch.


> PyTorch pinned commit is checkout and is built from source.

I think this is the biggest behavior change from users perspective, it's better to share for more feedback before land the merge. Basically, the impact to users is that until we have nightly branch created for ExecuTorch, users will have to build pytorch and domain libs from source when they build ExecuTorch repo from main branch. 

In general I agree to set up ET branches to be on-par with PyTorch where main branch will build from source, and nightly (off from stable branch, i.e. viable/strict) is for fast exploration w/o rebuilding from source.

**Comment by huydhn** [comment]
> In general I agree to set up ET branches to be on-par with PyTorch where main branch will build from source, and nightly (off from stable branch, i.e. viable/strict) is for fast exploration w/o rebuilding from source.

Yup, I think that would allow a good experience here for both power users who are actively developing PyTorch and ExecuTorch and need the latest trunk commits from both.  IMO, I don't think building PyTorch and domain libraries as a major blocker though given all the doc and guides people can find about the topic.  On the other hand, people who want to try out ExecuTorch could grab nightly branch which is tested against PyTorch and domain binaries **from the same date**.

File: .ci/docker/common/install_cache.sh [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+source "$(dirname "${BASH_SOURCE[0]}")/utils.sh"
+
+install_binary() {
+  echo "Downloading sccache binary from S3 repo"
+  curl --retry 3 https://s3.amazonaws.com/ossci-linux/sccache -o /opt/cache/bin/sccache
+  chmod +x /opt/cache/bin/sccache
+}
+
+mkdir -p /opt/cache/bin
+mkdir -p /opt/cache/lib
```
Comment (mergennachin) [inline .ci/docker/common/install_cache.sh:22]: do you need this directory?

File: .ci/docker/common/utils.sh [hunk 2]
```diff
@@ -27,3 +27,20 @@ conda_run() {
 pip_install() {
   as_ci_user conda run -n "py_${PYTHON_VERSION}" pip install --progress-bar off "$@"
 }
+
+init_sccache() {
```
Comment (mergennachin) [inline .ci/docker/common/utils.sh:31]: np: why in the utils.sh file? can it be just inside install_cache.sh file directory?

File: .ci/docker/common/install_pytorch.sh [hunk 3]
```diff
@@ -0,0 +1,50 @@
+#!/bin/bash
```
Comment (mergennachin) [inline .ci/docker/common/install_pytorch.sh:1]: lots of duplication code in .ci/scripts/utils.sh

is it possible to dedupe?

**Review by mergennachin — APPROVED** [review]
See inline

**Review by guangy10 — APPROVED** [review]
There are some linter errors to fix before merge, looks good to me otherwise

File: .ci/docker/common/install_pytorch.sh (hunk 3 above)
Comment (huydhn) [reply .ci/docker/common/install_pytorch.sh:1]: Yeah, I will follow up with another BE diff to dedup this. The underlying issue here is that Linux build uses those in `.ci/docker/common`, while MacOS build uses those under `.ci/scripts`.  There are subtle difference for example `as_ci_user`, so they are worth another PR

**Pull request closed without merging**